### ISSUE TITLE
device parsing with genie.libs.conf.device libs

### ISSUE THREAD

**Issue by sysp0** [CiscoTestAutomation/genielibs#126]
Hi,

i have problem when parse to another way device connect.

When I load the devices through the testbed using the code below, the "custom" variable is set correctly.

`custom : {"abstraction": {"order": ["os", "platform", "model"]}}:`

my testbed:
```yml
devices:
  device1:
    connections:
      cli:
        ip: 192.168.70.208
        protocol: ssh
    credentials:
      default:
        password: admin
        username: admin
      enable:
        password: admin
    os: ios
    type: ios
```

for example:
```python
from pyats.topology import loader
t = loader.load("testbed.yml")
device = t.devices["device1"]
device.connect()
device.parse("show ip interface brief") # OK
print(device.custom) # {"abstraction": {"order": ["os", "platform", "model"]}}
```
However, when I create the device object using the following code, the "custom" variable is not set and I get an error:

```python
from genie.libs.conf.device import Device
device = Device(
    "device",
    os="ios", type="ios",
    connections={"mgmt": {"protocol": "ssh", "ip": "192.168.70.208", "port": "22"}},
    credentials={
        "default": {"username": "admin", "password": "admin"},
        "enable": {"password": "admin"},
    },
)
device.connect()
device.parse("show ip interface brief") # raises an error
print(device.custom)  # AttrDict({})
```

The error message is :
```
Traceback (most recent call last):
  File "/home/w/pyats/device.py", line 50, in <module>
    device.parse("show ip interface brief")
  File "src/genie/conf/base/device.py", line 529, in genie.conf.base.device.Device.parse
  File "/home/w/pyats/venv/lib/python3.8/site-packages/genie/libs/parser/utils/common.py", line 257, in get_parser
    lookup = Lookup.from_device(device, packages={'parser': parser})
  File "/home/w/pyats/venv/lib/python3.8/site-packages/genie/abstract/magic.py", line 263, in from_device
    tokens = cls.tokens_from_device(device, default_tokens=default_tokens)
  File "/home/w/pyats/venv/lib/python3.8/site-packages/genie/abstract/magic.py", line 207, in tokens_from_device
    raise ValueError('Expected device to have custom.abstraction '
ValueError: Expected device to have custom.abstraction definition, or provide default tokens to use in Lookup.from_device() call

Process finished with exit code 1
```

I found that I can fix this issue by adding the "custom" variable to the Device class constructor, like so:

```python
device = Device(
    "device",
    os="ios", type="ios",
    connections={"mgmt": {"protocol": "ssh", "ip": "192.168.70.208", "port": "22"}},
    credentials={
        "default": {"username": "admin", "password": "admin"},
        "enable": {"password": "admin"},
    },
    custom={"abstraction": {"order": ["os", "platform", "model"]}},
)
device.connect()
device.parse("show ip interface brief") # OK
```

# Questions

- Could you please clarify whether this is expected behavior, or whether there is a way to populate the "custom" variable without having to do it manually?(when use `genie.libs.conf.device.Device` libs )
- If the initialization must be done manually, could you please provide a brief description of the variable(`custom`)  and why it should be assigned a value, as well as instructions on how to populate it correctly?
Thank you very much for your time and help.

# Version
```
You are currently running pyATS version: 23.2
Python: 3.8.10 [64bit]

  Package                      Version
  ---------------------------- -------
  genie                        23.2   
  genie.libs.clean             23.2   
  genie.libs.conf              23.2   
  genie.libs.filetransferutils 23.2   
  genie.libs.health            23.2   
  genie.libs.ops               23.2   
  genie.libs.parser            23.2   
  genie.libs.robot             23.2   
  genie.libs.sdk               23.2   
  genie.telemetry              23.2   
  genie.trafficgen             23.2   
  pyats                        23.2   
  pyats.aereport               23.2   
  pyats.aetest                 23.2   
  pyats.async                  23.2   
  pyats.connections            23.2   
  pyats.contrib                23.2   
  pyats.datastructures         23.2   
  pyats.easypy                 23.2   
  pyats.kleenex                23.2   
  pyats.log                    23.2   
  pyats.reporter               23.2   
  pyats.results                23.2   
  pyats.robot                  23.2   
  pyats.tcl                    23.2   
  pyats.topology               23.2   
  pyats.utils                  23.2   
  unicon                       23.2   
  unicon.plugins               23.2   
  yang.connector               23.2   

```

**Comment by sumitsehgal123** [comment]
Hi ,
Thanks for reaching out piestack support team,
Recommend is to always use the latest pyats version (i.e. Current released version 23.7)

**Comment by sumitsehgal123** [comment]
Expected Behavior and Populating the "custom" Variable:
Yes, the behavior you described is expected. The custom variable in the genie.libs.conf.device.Device class is intentionally left unpopulated by default. This means that if you want to use the custom variable, you need to explicitly provide values for it when creating a device object.

Description of the "custom" Variable and Its Purpose:
The custom variable is designed to be a user-defined attribute that allows you to extend the device object with additional metadata or configuration that might not be part of the standard attributes provided by the Genie library. It provides a way to tailor the behavior of the library to your specific needs.

The purpose of the custom variable is to give you the flexibility to add custom information or configuration that is relevant to your network environment but may not be covered by the predefined attributes of a device. This could include things like specific configuration templates, custom parsing logic, additional connection information, or any other metadata that you find useful.

The custom variable is a dictionary, so you can use it to store any key-value pairs that are relevant to your use case. The keys can be strings (representing the names of the custom attributes), and the values can be any appropriate data type (such as dictionaries, lists, strings, etc.).

Populating the "custom" Variable:
To populate the custom variable correctly, you can follow these steps:

a. When creating a device object using the Device class constructor, add the custom dictionary to the constructor and populate it with your desired custom attributes:

from genie.libs.conf.device import Device

device = Device(
    "device",
    os="ios", type="ios",
    connections={"mgmt": {"protocol": "ssh", "ip": "192.168.70.208", "port": "22"}},
    credentials={
        "default": {"username": "admin", "password": "admin"},
        "enable": {"password": "admin"},
    },
    custom={"my_custom_attribute": "custom_value"}
)
Access and use the custom attributes within your code:
`print(device.custom) # {"my_custom_attribute": "custom_value"}

`
By populating the custom variable, you can extend the device object's capabilities to suit your specific requirements. Remember that the custom variable is a powerful tool for customization and flexibility, allowing you to enhance the behavior of the Genie library based on your network environment and use cases.
